mitodl/open-edx-plugins
# feat: add ol_openedx_otel_monitoring plugin

# What are the relevant tickets?
https://github.com/mitodl/ol-infrastructure/issues/827

# Description (What does it do?)
This PR adds OpenTelemetry monitoring (traces) to our participating Django app (a.k.a. Plugin App) for better insights into our app's performance.

> ol_openedx_otel_monitoring

# How can this be tested?
Follow these steps in a terminal on your machine:

- Navigate to open-edx-plugins directory

- If you haven't done so already, run ./pants build

- Run ./pants package ::. This will create a "dist" directory inside "open-edx-plugins" directory with ".whl" & ".tar.gz" format packages for "edx_sysadmin" and other "ol_openedx_*" plugins in "open-edx-plugins/src")

- Move/copy any of the ".whl" or ".tar.gz" files for this plugin that were generated in the above step to the machine/container running Open edX (NOTE: If running devstack via Docker, you can use docker cp to copy these files into your LMS or CMS containers)

- Run a shell in the machine/container running Open edX, and install this plugin using pip

After installing the plugin, you can check if it's working properly by going to `<EDX_BASE_URL>/otel/healthcheck/` in your web browser. If the plugin is set up correctly, you should see trace data appearing in the console, provided that the console is set as the exporter.

# Additional Context

- Added tracing to capture info about requests and other operations.
- Set up middleware for future customizations.
- Added error handling for OpenTelemetry setup.
- Can choose different data exporters (e.g., console, otlp).
- New OTEL_CONFIG in settings to manage telemetry.

Added Dependencies:

```
opentelemetry-api
opentelemetry-sdk
opentelemetry-instrumentation-django
opentelemetry-exporter-richconsole
opentelemetry-exporter-otlp-proto-http
```


# Configuration

The plugin offers various settings for customization and functionality tuning:

`OTEL_CONFIGS.OTEL_ENABLED`: Control the activation of OpenTelemetry tracing through configuration settings.
`OTEL_CONFIGS.TRACES_EXPORTER`: Exporter Selection: Choose between different exporters like console and OTLP based on the environment (local or production).
`OTEL_CONFIGS.OTEL_INSTRUMENTATION_SQLCOMMENTER_ENABLED`: Option to enable or disable SQLCommenter from the configuration.

- Specific Settings for OTLPHTTP Exporter
`OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`: Specify the endpoint URL for sending trace data.
`OTEL_EXPORTER_OTLP_TRACES_HEADERS`: Required for authorization, format - `'{"Authorization": "Basic <base64_encoded_string>"}'`
`OTEL_EXPORTER_OTLP_TRACES_TIMEOUT`: Set the timeout for accessing the endpoint.

## Checklist:
- [x] Add more detailed documentation later.
